src/imitation/algorithms/sqil.py

```diff
@@ -0,0 +1,321 @@
+                at the creation of the instance.
+
+
+        """
+        self.venv = venv
+
+        super().__init__(demonstrations=demonstrations, custom_logger=custom_logger)
+
+        self.orig_train_freq = train_freq
+
+        self.dqn = dqn.DQN(
+            policy=policy,
+            env=venv,
+            learning_rate=learning_rate,
+            buffer_size=buffer_size,
+            learning_starts=learning_starts,
+            batch_size=batch_size,
+            tau=tau,
+            gamma=gamma,
+            train_freq=train_freq,
+            gradient_steps=gradient_steps,
+            replay_buffer_class=replay_buffer_class,
+            replay_buffer_kwargs=replay_buffer_kwargs,
+            optimize_memory_usage=optimize_memory_usage,
+            target_update_interval=target_update_interval,
+            exploration_fraction=exploration_fraction,
+            exploration_initial_eps=exploration_initial_eps,
+            exploration_final_eps=exploration_final_eps,
+            max_grad_norm=max_grad_norm,
+            tensorboard_log=tensorboard_log,
+            policy_kwargs=policy_kwargs,
+            verbose=verbose,
+            seed=seed,
+            device=device,
+            _init_setup_model=_init_setup_model,
+        )
+
+    def set_demonstrations(self, demonstrations: AnyTransitions) -> None:
+        # If demonstrations is a list of trajectories,
+        # flatten it into a list of transitions
+        if isinstance(demonstrations, Iterable):
+            item, demonstrations = get_first_iter_element(  # type: ignore[assignment]
+                demonstrations,  # type: ignore[assignment]
+            )
+            if isinstance(item, types.Trajectory):
+                demonstrations = flatten_trajectories(
+                    demonstrations,  # type: ignore[arg-type]
+                )
+
+        n_samples = len(demonstrations)  # type: ignore[arg-type]
+        self.expert_buffer = ReplayBuffer(
+            n_samples,
+            self.venv.observation_space,
+            self.venv.action_space,
+            handle_timeout_termination=False,
+        )
+
+        for transition in demonstrations:
+            self.expert_buffer.add(
+                obs=np.array(transition["obs"]),  # type: ignore[index]
+                next_obs=np.array(transition["next_obs"]),  # type: ignore[index]
+                action=np.array(transition["acts"]),  # type: ignore[index]
+                done=np.array(transition["dones"]),  # type: ignore[index]
+                reward=np.array(1),
+                infos=[{}],
+            )
+
+    def train(self, *, total_timesteps: int):
+        self.learn_dqn(total_timesteps=total_timesteps)
+
+    @property
+    def policy(self) -> policies.BasePolicy:
+        assert isinstance(self.dqn.policy, policies.BasePolicy)
+        return self.dqn.policy
+
+    def train_dqn(self, gradient_steps: int, batch_size: int = 100) -> None:
```

Comment:
Bit of a shame we need to duplicate so much from DQN implementation (I assume?), but probably unavoidable